### ISSUE TITLE
Environment variable substitution stopped working.

### ISSUE THREAD

**Issue by mgoodings** [buildkite/agent#647]
I cannot get environment variable substitution to work inside of the environment block at the top of the pipeline.yml after upgrading my Agent.

At the step level it seems to be working fine however it's necessary for me to use it at the environment block so all subsequent steps have the environment variables applied to them on execution.

This is using Agent v3.0-beta.38

![screen shot 2018-02-06 at 1 52 08 pm](https://user-images.githubusercontent.com/1767585/35839524-f8386012-0b44-11e8-9272-e715291d9dc9.png)

The same pipeline.yml running on Agent v3.0-beta.19

![screen shot 2018-02-06 at 1 54 12 pm](https://user-images.githubusercontent.com/1767585/35839582-485b2598-0b45-11e8-8046-9d87cbf687b4.png)

```
env:
  ACCOUNT: "${ACCOUNT?is not set. Please specify an account}"
  ENVIRONMENT: "${ENVIRONMENT?is not set. Please specify an environment}"
  BUILD_NUMBER: "${BUILDKITE_BUILD_NUMBER}"

steps:
  - name: ":docker: Build ${ENVIRONMENT}"
    command: ".buildkite/steps/build"
```

**Comment by lox** [comment]
Hey, sorry you're having problems @mgoodings. Let's figure out what is happening here. Do you happen to know what agent you were on before?

**Comment by lox** [comment]
Oh I just saw 🤦🏼‍♂️

**Comment by lox** [comment]
We changed our interpolation code in Beta 35. Let me investigate.

**Comment by lox** [comment]
Is your expectation that the `${ENVIRONMENT}` in the steps would be the environment from when the pipeline uploaded? If `$ACCOUNT` and `$ENVIRONMENT` weren't available in your pipeline upload step, would you expect the pipeline upload step to fail?

**Comment by lox** [comment]
Just to be sure, are you positive that a v3 agent handled the pipeline upload step?

**Comment by mgoodings** [comment]
I would expect the pipeline to fail, that's what it did before. I posted the versions of the agents being used.

**Comment by mgoodings** [comment]
Any progress on this?

Seem's pretty clear that:

```
env:
  ENVIRONMENT: "${ENVIRONMENT?is not set. Please specify an environment}"
```

Is not being interpolated at all and just being read as a raw string.

Then:

```
steps:
  - name: ":docker: Build ${ENVIRONMENT}"
```

Is being interpolated correctly and putting the value of $ENVIRONMENT which is now the raw string from the env block.

**Comment by lox** [comment]
I started looking into this and added a series of integration tests with exactly those cases and it behaved as expected, so I'm back to the drawing board at present.

When I create a pipeline with the yaml that you provided, this is what I get: 
![image](https://user-images.githubusercontent.com/15758/36135673-c7317aa2-10e0-11e8-8bcd-8fda48169d6f.png)

Could you drop us a support email with a pipeline that this is failing in so I can investigate? Could you also triple check that the pipeline upload step is running on an Agent v3? This would be exactly the behaviour that you would see if you used a v2 agent to do the pipeline upload.

**Comment by lox** [comment]
For reference, here are the interpolation test cases:

https://github.com/buildkite/interpolate/blob/master/interpolate_test.go#L241-L243

**Comment by mgoodings** [comment]
This is using an out of the box Elastic CI Stack v2.3.3 setup a week ago.

I just ran this on the agent today. Error shown and agent.

![screen shot 2018-02-13 at 5 23 39 pm](https://user-images.githubusercontent.com/1767585/36136091-d12ccba4-10e2-11e8-8c5f-1b3f0f864ed8.png)

![screen shot 2018-02-13 at 5 23 50 pm](https://user-images.githubusercontent.com/1767585/36136095-d5c93792-10e2-11e8-816f-c5a57e9c7af0.png)

Entire Pipeline YML:

```
env:
  ACCOUNT: "${ACCOUNT?is not set. Please specify an account}"
  PROJECT_SLUG: "${BUILDKITE_PIPELINE_SLUG}"
  BUILD_NUMBER: "${BUILDKITE_BUILD_NUMBER}"

steps:
  - block: ":rocket: Deploy to ${ACCOUNT}"

  - name: ":cloudformation: Deploy ${ACCOUNT}"
    command: .buildkite/steps/deploy
    agents:
      queue: bluesky
    plugins:
      irexchange/aws-assume-role#v0.1.0:
        role: "${DEPLOY_ROLE}"
      docker-compose#v1.8.2:
        run: deployer
```

**Comment by mgoodings** [comment]
Issue is from here:

https://github.com/buildkite/agent/blob/master/agent/pipeline_parser.go#L42

and here:

https://github.com/buildkite/agent/blob/master/agent/pipeline_parser.go#L62

It passes the env block multiple times, first using the raw values from the block then interpolating.

Because the output variable is the same as the input variable, it overwrites the input before the second pass can occur using the actual process environment block. So the second pass just interpolates the string value of the first.

I use this env block to set the environment on all subsequent steps. Not sure what your acceptable solution is.

**Comment by mgoodings** [comment]
Changed YML above to show first/second pass issue from using the same variable name.

Using a different variable name it works as intended. We have common variables across all of our projects and I dislike the idea of a hacky fix like this.

```
env:
  ACCOUNT: "${IRX_ACCOUNT?is not set. Please specify an account}"
```

![screen shot 2018-02-13 at 6 04 41 pm](https://user-images.githubusercontent.com/1767585/36137243-6b677642-10e8-11e8-8bec-9ab9fdc08cee.png)

**Comment by lox** [comment]
Ah yup, classic case of missing the forest for the trees :) Sorry about that @mgoodings, it's been a hectic few days and I've been multi-tasking too much. 

We pre-parse the top-level env block to allow it to be interpolated later into the steps. See https://github.com/buildkite/agent/issues/544 for backstory there. The difficulty there is that order of maps in golang is undefined (technically it is as well in json), so processing the ENV block in order for self-referential variables is really difficult, as we get them back in a random order. We compromise with the current algorithm of two-passes of ENV passing, which has the downside of the issue you are reporting (which didn't occur to me).

The only solution I can see to this is either:
* Figure out how to get consistent ordering in the YAML/JSON parsing and then process ENV block in order
* Come up with a way to extract dependencies of ENV and process them in dependency order

Neither are trivial unfortunately. Any other ideas?

**Comment by mgoodings** [comment]
I understand the rational for the change but is this a case of the env block trying to do too much?

The env block is currently setting environment variables (the new functionality) that can be used in and by the pipeline.yml but also applying those environment variables to each step as well (the old functionality).

I feel like those are two distinct things, one is for inputs (generating steps), the other is for outputs (setting up step environment).

Could we possibly come up with the concept of environment variable forwarding to steps as that's all I'm really trying to do. So I don't have to specify the env for every step and my common variables could be forwarded.

For example we have the following variables expected by most projects:
* ACCOUNT
* ENVIRONMENT
* PROJECT_SLUG
* BUILD_NUMBER

We have intentionally avoided using BUILDKITE_* variables here to allow local execution to be no different than running in Buildkite. Set those environment variables, docker-compose and job done.

**Comment by lox** [comment]
To me that sounds like you want an `environment` hook that checks for those env? I'm confused by what you mean by "environment variable forwarding". You don't have to explicitly specify environment in the pipeline for steps, it's there implicitly based on whatever environment your agent picks up from the host, or variables exported in the hooks. 

The other complexity here is WHEN things are interpolated / evaluated. Variable interpolation happens on pipeline upload, which means they have to be set for whatever is running the pipeline upload. Folks often get confused with this and assume they are interpolated at step run time.

**Comment by mgoodings** [comment]
I feel like my above comment was misinterpreted around the complexity introduced by the current substitution process. The env block is fulfilling two needs now, one is to make environment variables available to the steps being generated in the pipeline.yml itself, the other is to provide predefined environment variables available for use when the step actually executes.

Previously it didn't care about making environment variables available to the pipeline.yml it just set the environment block for all defined steps.

Any environment variables I set against the pipeline upload step in the actual web interface won't be automatically forwarded to the steps created by that pipeline unless I put them in the environment block. So are you saying I should be using a global hook across all agents to do this?

**Comment by lox** [comment]
> I feel like my above comment was misinterpreted around the complexity introduced by the current substitution process.

Apologies if so! 

> Previously it didn't care about making environment variables available to the pipeline.yml it just set the environment block for all defined steps.

Yup, and I agree it's possible we are asking the `ENV` block to do too much with this change. It's tricky, in that lots of folks expect variables defined in that block to be available for interpolation. Perhaps we haven't found the right balance. 

> Any environment variables I set against the pipeline upload step in the actual web interface won't be automatically forwarded to the steps created by that pipeline unless I put them in the environment block. So are you saying I should be using a global hook across all agents to do this?

Yup, I think an `environment` hook in all of your agents would be how I would handle this. Might that solve your problem?

**Comment by mgoodings** [comment]
So is it possible to use an environment hook to take a variable defined in the web interface during a particular pipeline upload step and pass that variable to any steps created during that pipeline upload only?

In the above example that's the scenario I was outlining. I can currently make it work by using a different variable name in the pipeline upload step definition and interpolating to the correct one from the environment block so that all steps resulting from that pipeline upload have the correct environment variable set.

**Comment by lox** [comment]
@mgoodings Ah right, I think I see where you are coming from. I rarely have anything other than a single `buildkite-agent pipeline upload` configured via the web interface. The environment textarea in the pipeline configuration should apply to all the steps though, right? So you shouldn't need a mechanism to propagate them from pipeline upload to the other steps, they will just be available to all steps without you having to do anything.

FWIW, if you are hoping to propagate key/vals between steps/jobs/builds, we recommend [Build Metadata](https://buildkite.com/docs/builds/build-meta-data)

**Comment by mgoodings** [comment]
If that's how it worked it would solve all my issues, but sadly not.

Here's an example:

![screen shot 2018-02-14 at 1 35 59 pm](https://user-images.githubusercontent.com/1767585/36185246-2b6ad526-118c-11e8-99dc-e43d4528a5f8.png)

If I don't set the ACCOUNT variable in the env block of the yml the variable is not set in the generated step:

![screen shot 2018-02-14 at 1 37 40 pm](https://user-images.githubusercontent.com/1767585/36185262-49292fae-118c-11e8-8eb6-70b8e6736899.png)

I know all of this could be solved by using a shell script to pragmatically generate a fixed build pipeline from a single upload step, however this pipeline.yml can be used for many environments and we can add/remove environments on the fly with no code changes and no complex templated shell script with this kind of setup.

If I'm just going far too left field with this please let me know. I appreciate all of your help so far.

**Comment by lox** [comment]
Ah right, I was thinking you were using the top-level env block that applies to all steps. You are correct, if you want different values for `ACCOUNT` then you will need to propagate them as you have been trying to do. I've got a clear idea of what the issue is and the fix for it (basically evaluating the top level ENV block in dependency order). We'll endeavour to get a fix out in the next beta. Thanks for sticking with explaining it to me @mgoodings!

**Comment by lox** [comment]
Fixed in #668
